checks/src/main/kotlin/com/faithlife/lint/FiniteWhenCasesDetector.kt

```diff
@@ -0,0 +1,77 @@
+package com.faithlife.lint
+
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Incident
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import org.jetbrains.uast.UExpression
+import org.jetbrains.uast.kotlin.KotlinUSwitchEntry
+import org.jetbrains.uast.kotlin.KotlinUSwitchExpression
+
+@Suppress("UnstableApiUsage")
+class FiniteWhenCasesDetector : Detector(), SourceCodeScanner {
+    override fun getApplicableUastTypes() = listOf(UExpression::class.java)
+
+    override fun createUastHandler(context: JavaContext) = object : UElementHandler() {
+        override fun visitExpression(node: UExpression) {
+            val whenExpression = node as? KotlinUSwitchExpression ?: return
+            val subjectType = whenExpression.expression?.getExpressionType() ?: return
+
+            val subjectTypeClass = context.evaluator.getTypeClass(subjectType)
+            val isSealed = context.evaluator.isSealed(subjectTypeClass)
+            val isEnum = context.evaluator.extendsClass(subjectTypeClass, "java.lang.Enum", true)
+
+            val elseBranch = node.body.expressions
+                .filterIsInstance<KotlinUSwitchEntry>()
+                .find { it.caseValues.isEmpty() } // else -> is the only possible case with empty values
+
+            if ((isEnum || isSealed) && elseBranch != null) {
+                val incident = Incident(context)
+                    .issue(ISSUE)
+                    .message(MESSAGE)
+                    .scope(node)
+                    .location(context.getLocation(elseBranch))
+
+                incident.report()
+            }
+        }
+    }
+
+    companion object {
+        private const val MESSAGE = "Prefer explicit case handling over else."
+
+        val ISSUE = Issue.create(
+            "FiniteWhenCasesDetector",
+            "Avoid else where possible",
+            """
+                $MESSAGE
+
+                Explicit handling of when cases encourages reconsideration of
+                relevant when expressions if the subject of the expression changes.
+
+                This should not be used if the when subject might possibly introduce
+                new cases without this module being recompiled. One common way this can
+                happen is if the application consumes jars delivered out of sync with the app.
+
+                This usually doesn't happen with android applications, which typically
+                consume all libraries being used (other than the Android Framework)
+                at compile time.
```

Comment:
```suggestion
                This should not be used if the when subject might possibly introduce
                new cases without this module being recompiled. One common way this can
                happen is if the application consumes jars delivered out of sync with the app.

                This usually doesn't happen with android applications, which typically
                consume all libraries being used (other than the Android Framework)
                at compile time.

                This check should only run in modules that will be compiled with an app.
                It should not run in library code since library dependencies may change without recompiling the library.
```

@ddunkin, is this a helpful clarification?

There's still some danger in a when subject being defined in the android framework. I'll probably look into not enforcing this in that case.

Reply:
Yeah, that's helpful.

Can you configure rules? It still could be useful to apply to non-public subject types because those shouldn't ever change without recompilation.

Reply:
It's possible to change the severity of the issue and ignore paths in a config file. Suppressions are possible in code, too.

We have a Gradle plugin that's applied to all our apps that can enable the check for relevant modules. So if the check remains opt-in, all of our app modules could be opted in by default.

>  It still could be useful to apply to non-public subject

Good idea.